contracts/Minter.sol

```diff
@@ -97,35 +103,68 @@ contract SPSMinter {
       unchecked {
         totalMinted = totalMinted + amount;
       }
-      token.mint(pools[i].receiver, amount);
 
-      emit Mint(pools[i].receiver, amount);
+      if (amount > 0){
+        token.mint(pools[i].receiver, amount);
+        emit Mint(pools[i].receiver, amount);
+      }
+
+      unchecked { ++i; }
     }
   }
 
   /**
    * @notice Add new pool, can be called by admin
    * @param newReceiver Address of the receiver
    * @param newAmount Amount of tokens per block
+   * @param newReductionBlocks Number of blocks between emission reduction
+   * @param newReductionBps Number of basis points to reduce emission
    */
-  function addPool(address newReceiver, uint256 newAmount) external onlyAdmin {
+  function addPool(address newReceiver, uint256 newAmount, uint256 newReductionBlocks, uint256 newReductionBps) external onlyAdmin {
     require(pools.length < poolsCap, 'SPSMinter: Pools cap reached');
     require(newAmount <= maxToPoolPerBlock, 'SPSMinter: Maximum amount per block reached');
-    pools.push(Pool(newReceiver, newAmount));
-    emit PoolAdded(newReceiver, newAmount);
+    require(newReductionBps <= 10000, "SPSMinter: newReductionBps cannot be larger than 10000");
+    pools.push(Pool(newReceiver, newAmount, newReductionBlocks, newReductionBps, block.number));
+    emit PoolAdded(newReceiver, newAmount, newReductionBlocks, newReductionBps, block.number);
   }
 
   /**
    * @notice Update pool, can be called by admin
    * @param index Index in the array of the pool
    * @param newReceiver Address of the receiver
    * @param newAmount Amount of tokens per block
+   * @param newReductionBlocks Number of blocks between emission reduction
+   * @param newReductionBps Number of basis points (1 bps = 1/100th of 1%)to reduce emission
    */
-  function updatePool(uint256 index, address newReceiver, uint256 newAmount) external onlyAdmin {
+  function updatePool(uint256 index, address newReceiver, uint256 newAmount, uint256 newReductionBlocks, uint256 newReductionBps) external onlyAdmin {
     require(newAmount <= maxToPoolPerBlock, 'SPSMinter: Maximum amount per block reached');
+    require(newReductionBps <= 10000, "SPSMinter: newReductionBps cannot be larger than 10000");
     mint();
-    pools[index] = Pool(newReceiver, newAmount);
-    emit PoolUpdated(index, newReceiver, newAmount);
+    pools[index] = Pool(newReceiver, newAmount, newReductionBlocks, newReductionBps, block.number);
+    emit PoolUpdated(index, newReceiver, newAmount, newReductionBlocks, newReductionBps, block.number);
+  }
+
+  /**
+   * @notice Update emissions for one pool
+   * @param index Index in the array of the pool
+   */
+  function updateEmissions(uint256 index) public {
```

Comment:
Don't know why this didn't occur to me earlier, but I think we can remove a lot of complexity and gas cost by calculating the current emission rather than actually updating the state like this. If you replace `lastUpdate` in the Pool struct with something like `baseBlock`:

```solidity
  struct Pool {
    address receiver;
    uint256 amountPerBlock;
    uint256 reductionBlocks;
    uint256 reductionBps;
    uint256 baseBlock;
  }
```

Then you should be able to do something like this (untested):

```solidity
function getPoolEmission(uint256 index) public view returns (uint256) {
    uint256 _blocksElapsed = block.number - pools[index].baseBlock;
    uint256 _reductions = _blocksElapsed / pools[index].reductionBlocks;
    uint256 _reductionBps = _reductions * pools[index].reductionBps;
    return (pools[index].amountPerBlock * (BPS - _reductionBps)) / BPS;
 }
```

Inside the mint function you can then use the getter above to calculate the amount

```solidity
uint256 amount = getPoolEmission(i) * mintDifference;
```

Reply:
I was thinking the same, but the requirements I got were to reduce the amount using exponential decay (`...it would drop to 0.99 SPS / block, and then the next day it would drop to 0.9801 (1% reduction from 0.99), then 0.9702, etc)`, I tried using `P *(1 - r)**t` formula, but I didn't manage to get it working in Solidity (since 1-r is smaller than 1).

Reply:
I was thinking about the same approach, but the requirements I was given were to use exponential decay (`after the first day (28800 blocks) it would drop to 0.99 SPS / block, and then the next day it would drop to 0.9801 (1% reduction from 0.99), then 0.9702, etc`), I tried using it, but standard `x*(1-y)**z)` formula doesn't work in solidity (and I couldn't get it working with any other way either).

Reply:
Ahh yes, great point! My version is not exponential. Missed that.

>  I tried implementing it, but the standard x*(1-y)**z) formula doesn't work in solidity.

Just out of curiosity, what kind of issue did you hit implementing this formula?

If it's not possible, we should still be able to do something iteratively. Here's a quick attempt (not tested):

```solidity
  function getPoolEmission(uint256 index) public view returns (uint256) {
    uint256 _blocksElapsed = block.number - pools[index].baseBlock;
    uint256 _reductions = _blocksElapsed / pools[index].reductionBlocks;

    uint256 _reductionBps = _reductions * pools[index].reductionBps;
    uint256 _emission = pools[index].amountPerBlock;

    for (uint256 i = 0; i < _reductions; i++) {
      _emission = _emission * (BPS - _reductionBps) / BPS;

      if (minimumPayout > _emission) {
        return 0;
      }
    }

    return _emission;
  }

```

Reply:
The issue with using `x*(1-y)**z) formula ` was that `1-z`  should be less than 1, which doesn't work with solidity. 

I was thinking about an iterative approach but wasn't sure if gas savings are worth it, did some testing now, and while it's cheaper (~1k gas), I don't think it's worth changing it considering it will be used on BSC anyway, so gas cost is not an issue. 

While using a non-iterative approach requires calling `updateEmissions` every `reductionBlocks`, minter will be integrated with a yield farming contract where users will call it, so I don't think that will be an issue either.